src/program/graph.rs

```diff
@@ -233,24 +301,47 @@ impl InstructionBlock {
                     Ok(())
                 }
                 InstructionRole::RFControl => {
-                    let used_frames = program
+                    let used_frames: HashSet<&FrameIdentifier> = program
                         .get_frames_for_instruction(instruction, false)
-                        .unwrap_or_default();
-                    let blocked_frames = program
+                        .unwrap_or_default()
+                        .into_iter()
+                        .collect();
+                    let blocked_frames: HashSet<&FrameIdentifier> = program
                         .get_frames_for_instruction(instruction, true)
-                        .unwrap_or_default();
+                        .unwrap_or_default()
+                        .into_iter()
+                        .filter(|f| !used_frames.contains(f))
+                        .collect();
 
-                    // Take a dependency on any previous instructions to _block_ a frame which this instruction _uses_.
                     for frame in used_frames {
-                        let previous_node_id = last_instruction_by_frame
-                            .get(frame)
-                            .unwrap_or(&ScheduledGraphNode::BlockStart);
-                        add_dependency!(graph, *previous_node_id => node, ExecutionDependency::ReferenceFrame);
+                        let previous_node_ids = last_instruction_by_frame
+                            .entry(frame.clone())
+                            .or_insert(PreviousNodes {
+                                using: None,
+                                blocking: vec![ScheduledGraphNode::BlockStart]
+                                    .into_iter()
+                                    .collect(),
+                            })
```

Comment:
```rust
.or_default()
```

Reply:
Ah right! I added the default impl then forgot to use it